packages/txservice/src/transaction.ts

```diff
@@ -0,0 +1,216 @@
+import { ChainError } from "./error";
+import { ChainRpcProvider } from "./provider";
+import { FullTransaction, MinimalTransaction } from "./types";
+
+export class Transaction {
+
+  public responses: providers.TransactionResponse[] = [];
+  public receipt?: providers.TransactionReceipt;
+  public nonce?: number;
+  public nonceExpired = false;
+
+  // TODO: Not a fan of this arrangement, this code could be tightened up a bit.
+  public get gasPrice(): Promise<BigNumber> {
+    return this.getGasPrice();
+  }
+  private _gasPrice?: BigNumber;
+
+  public get fullTransaction(): Promise<FullTransaction> {
+    return new Promise(async (resolve) => 
+      resolve({
+        ...this.minTx,
+        gasPrice: await this.gasPrice,
+        nonce: this.nonce,
+      })
+    );
+  }
+
+  constructor(
+    private readonly log: BaseLogger,
+    private readonly provider: ChainRpcProvider,
+    private readonly minTx: MinimalTransaction,
+    private readonly config: TransactionServiceConfig,
+    initialGasPrice?: BigNumber,
+  ) {
+    this._gasPrice = initialGasPrice;
+  }
+
+  public async send() {
+    const method = this.send.name;
+
+    // Sanity check to make sure nonce is not expired.
+    if (this.nonceExpired) {
+      throw new ChainError(ChainError.reasons.NonceExpired, { method });
+    }
+    
+    const fullTransaction = await this.fullTransaction;
+    this.log.info(
+      {
+        method,
+        ...fullTransaction,
+        gasPrice: fullTransaction.gasPrice.toString(),
+      },
+      "Attempting to send transaction.",
+    );
+    const { response: _response, success } = await this.provider.sendTransaction(await this.fullTransaction);
+    if (!success) {
+      const error = _response as Error;
+      if (
+        this.responses.length >= 1 &&
+        (error.message.includes("nonce has already been used") ||
+          // If we get a 'nonce is too low' message, a previous tx has been mined, and ethers thought
+          // we were making another tx attempt with the same nonce.
+          error.message.includes("Transaction nonce is too low.") ||
+          // Another ethers message that we could potentially be getting back.
+          error.message.includes("There is another transaction with same nonce in the queue."))
+      ) {
+        this.nonceExpired = true;
+        this.log.info(
+          {
+            method,
+            ...fullTransaction,
+            gasPrice: fullTransaction.gasPrice.toString(),
+            error: error.message,
+          },
+          "Nonce already used.",
+        );
+      } else {
+        this.log.error({ method, error }, "Failed to send tx");
+        throw _response;
+      }
+    }
+    const response = _response as providers.TransactionResponse;
+
+    // Save nonce if applicable.
+    if (this.nonce === undefined) {
+      this.nonce = response.nonce;
+    }
+    // TODO: Should we be checking for wild case where this.nonce !== response.nonce?
```

Comment:
at the very least we should log it and everything related to this


cus that would be wild

Reply:
Yeah gonna add a log.warning I guess